### DISCUSSION TITLE
Avalonia and Silk.NET integration

### DISCUSSION THREAD

**Discussion by orosbogdan** [AvaloniaUI/Avalonia#8039] [Q&A]
Hello,
I'd like to use Avalonia for 2D and Silk.NET for 3D rendering together.
I have a few questions:
1. Does Avalonia offer rendering to a surface on a higher framerate than 60 fps ? 
2. Does Avalonia offer any possibility to be integrated easily with Silk.NET ? If yes, any sample ?

**Comment by martinrhan** [comment]
I am having same question

**Comment by kekekeks** [comment]
https://github.com/kekekeks/Avalonia-Silk.NET-Example/tree/master/AvaloniaSilkExample

**Reply by BenTec3d** [reply]
> https://github.com/kekekeks/Avalonia-Silk.NET-Example/tree/master/AvaloniaSilkExample

I've been using your example to follow an OpenGL tutorial inside of a multiplatform Avalonia project. I highly recommend anyone interested in using Silk.net inside of Avalonia to give it a look. Thanks a lot 👍🏻 

So far I only got it running on PC though. Would you happen to know how to get it to display OpenGL on Andoid as well? Does the OpenGlControlBase even support mobile platforms yet?

**Reply by kekekeks** [reply]
It does support Android, however you'll need to use OpenGL ES wrapper from Silk.NET.OpenGLES package instead of Silk.NET.OpenGL.

**Reply by BenTec3d** [reply]
I tried using Silk.NET.OpenGLES but it made no difference. That's why I assumed the OpenGlControlBase inside Avalonia is simply not yet supported on mobile.

I created a fresh project from the cross platform template and added a very basic HelloTriangle.  On desktop it works but on Android all I get is a blank screen with the background color set in MainView.
https://github.com/BenTec3d/AvaloniaSilkOpenGLES

If you or someone else could take a look and tell me what I need to change for HelloTriangle to be displayed on mobile that would be great. I want to pitch Avalonia UI + Silk.NET to my employer for a future project based on both being cross platform, which is the biggest problem with our current solution, so having the demo I am working on run on Android for the pitch would be amazing.

**Reply by anose001** [reply]
@BenTec3d, did you eventually get the triangle to work on mobile?

**Reply by orosbogdan** [reply]
@BenTec3d  @anose001 is it working on mobile?

**Comment by damian-666** [comment]
thanks kekekek  i was tooling for this might have to put an openTk window in there..  
@orosbogdan 

   re: > 60 ms..  rendering.  AFAIK    there is a way, 
   
     [System.Runtime.InteropServices.DllImport("winmm.dll", EntryPoint = "timeBeginPeriod")]
    public static extern uint timeBeginPeriod(uint uMilliseconds);
    
    set to 4  or so and try.
   
   but  the default timer resolution in windows is 1/60 or 16 ms..  that affects many things waithandles , timers, rely on that.   if you use a high-resolution  timer  around sleep(1) you will get results that vary from 1 to 16 ms.    and then locks and timers are useless if going for > 60 for smoothing framerates with physics or whatever ..    if you need to poll the touch you might want to do it faster, or poll keys    just display  last frame every  60 hz ( fast enough usually)   , or that might be timed with the vsync of the screen naysays


if you need faster physics or more tight processing of animation frames, run a background thread and to even the frame rate you can meaure the time it takes.  and use a lockless producer consumer pattern ( copy the view stuff to a read buffer and write the next frame and swap )  .   i can run a interactive game , capped at about 300 fps .  try not to use locks., copy is super fast in net 8  especially. Simd is autovectorized. 

you can set this to1    but it will drain your battery and use more power just for your tool..  it will not affect the other apps and you don't really have to set it back.. you can.. with endperiod.   maybe on idle.. for laptops,   it will go back to normal on app close.
    

  for non windows linux, ios,  i don't know, there's a similar low level way and android, if you can draw faster id like to konw i havent tried it.  there is  a way if you google aroudn..   i would start by timing sleep (1) in anycase.

**Reply by maxkatz6** [reply]
On windows WinUI/Direct Composition timer is used by default which can go way higher than 60fps.
